server/src/ingestion.rs

```diff
@@ -0,0 +1,135 @@
+/* Copyright 2022 The MiniModelarDB Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! Support for data ingestion from an MQTT broker.
+//!
+//! To use the ingestor, run the "ingestor.start()" method. This method first creates the client. Then
+//! it uses the created client to connect to the broker and subscribe to the specified topics. The
+//! connection message stream is then looped over to ingest the messages that are published to the topics.
+
+use std::{process, time::Duration};
+
+use futures::{executor::block_on, stream::StreamExt};
+use paho_mqtt as mqtt;
+use paho_mqtt::{AsyncClient, AsyncReceiver, Message};
+
+use crate::storage::StorageEngine;
+
+/// A single MQTT client that can subscribe to `broker` and ingest messages from `topics`. Note that
+/// after creation, the ingestor needs to be started to ingest messages.
+pub struct Ingestor {
+    /// Server URI for the MQTT broker.
+    broker: &'static str,
+    /// ID that is used to uniquely identify the ingestor as a client to the MQTT broker.
+    client_id: &'static str,
+    /// Specific topics that should be subscribed to. Use "\[*]" to subscribe to all topics.
+    topics: &'static [&'static str],
+    /// The quality of service for each subscribed-to topic. Should be same length as `topics` field.
+    qos: &'static [i32],
+}
+
+impl Ingestor {
+    pub fn new(
```

Comment:
Maybe redefine the method to `try_new(...) -> Result<Self, Error or String>` where `Self` is only returned if `topics` and `qos` is the same size, and if possible, the `paho_mqtt` client can connect.

Reply:
Changed to `try_new` where `Self` is only returned if `topics` and `qos` is of the same length. Also added some tests for this.